### PR TITLE
Decompile unk_020851B8

### PR DESCRIPTION
**Pull request by Gogume1er**
First PR, I don't know what I'm doing

### CODE REVIEW

File: arm9/src/unk_020851B8.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    ret->unk00[0x2] = 0;
+    ret->unk00[0x3] = 0;
+    ret->unk00[0x4] = 0;
+    ret->unk00[0x5] = 0;
+    ret->unk00[0x6] = 0;
+    ret->unk00[0x7] = 0;
+    ret->unk00[0x8] = 0;
+    ret->unk00[0x9] = 0;
+    ret->unk00[0xA] = 0;
+    ret->unk00[0xB] = 0;
```
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:19]: You sure it's an array access and not individual struct fields?

File: arm9/src/unk_020851B8.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    param0->unk00[0xA] = param2;
+    param0->unk00[0xB] = param3;
+}
+
+THUMB_FUNC void FUN_02085200(u32 *param0, u16 param1)
+{
+    u32 index = param1 / 32;
+    u32 tag = 1 << (param1 % 32);
+    u32 value = *(param0 + index);
+    *(param0 + index) = value | tag;
```
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:47]: This is mega amounts of eww. param0 is probably of type `struct UnkStruct_020851B8 *` and its first field is a u32[2].

File: arm9/src/unk_020851B8.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    u32 r2 = 1;
+    u32 index = param1 / 32;
+    u32 tag = 1 << (param1 % 32);
+
+    if (*(param0 + index) & tag)
+    {
+        r2 = 0;
+    }
+
+    return (r2 << 0x18) >> 0x18;
```
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:60]: This indicates that the return type is u8.

File: include/unk_020851B8.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#ifndef POKEDIAMOND_UNK_020851B8_H
+#define POKEDIAMOND_UNK_020851B8_H
+
+struct UnkStruct_020851B8 {
+    u8 unk00[0xC];
+};
```
Comment (PikalaxALT) [inline include/unk_020851B8.h:6]: More likely:
```c
{
  u32 flags[2];
  u8 unk_8;
  u8 unk_9;
  u8 unk_A;
  u8 unk_B;
};
```

File: arm9/src/unk_020851B8.c (hunk 1 above)
Comment (PikalaxALT) [reply arm9/src/unk_020851B8.c:19]: This matches: 
```c
static inline void _clear(u8 * ptr) {
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
    *ptr++ = 0;
}

struct UnkStruct_020851B8 * FUN_020851B8(u32 heap_id)
{
    struct UnkStruct_020851B8 * ret = AllocFromHeap(heap_id, sizeof(struct UnkStruct_020851B8));
    _clear((u8 *)ret);
    return ret;
}
```

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

File: arm9/src/unk_020851B8.c (hunk 3 above)
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:60]: Probably a ternary expression here.

File: include/unk_020851B8.h (hunk 4 above)
Comment (Gogume1er) [reply include/unk_020851B8.h:6]: Seems to match with the 64 berries, I didn't see that, mb

File: arm9/src/unk_020851B8.c [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    param0->unk_9 = param1;
+    param0->unk_A = param2;
+    param0->unk_B = param3;
+}
+
+THUMB_FUNC void FUN_02085200(struct UnkStruct_020851B8 *param0, u16 berry_id)
+{
+    u32 index = berry_id / 32;
+    u32 tag = 1 << (berry_id % 32);
+    u32 value = param0->flags[index];
```
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:50]: Make this an in-place operation (`|=`)

File: arm9/src/unk_020851B8.c [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+{
+    param0->unk_9 = param1;
+    param0->unk_A = param2;
+    param0->unk_B = param3;
+}
+
+THUMB_FUNC void FUN_02085200(struct UnkStruct_020851B8 *param0, u16 berry_id)
+{
+    u32 index = berry_id / 32;
+    u32 tag = 1 << (berry_id % 32);
```
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:49]: Temp variables are not necessary, just inline the div/mod

File: arm9/src/unk_020851B8.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    param0->flags[index] |= tag;
+}
+
+THUMB_FUNC u8 FUN_02085224(struct UnkStruct_020851B8 *param0, u16 berry_id)
+{
+    u32 index = berry_id / 32;
+    u32 tag = 1 << (berry_id % 32);
```
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:56]: See comment on L48-L49

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

File: arm9/src/unk_020851B8.c [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    *ptr++ = 0;
+    *ptr++ = 0;
+}
+
+THUMB_FUNC struct UnkStruct_020851B8 *FUN_020851B8(u32 heap_id)
+{
+    struct UnkStruct_020851B8 *ptr = AllocFromHeap(heap_id, sizeof(struct UnkStruct_020851B8));
+    _clear((u8 *)ptr);
+    return ptr;
+}
```
Comment (PikalaxALT) [inline arm9/src/unk_020851B8.c:27]: Upon further investigation, this also matches with the metrowerks `__builtin__clear` macro:

```c
THUMB_FUNC struct UnkStruct_020851B8 *FUN_020851B8(u32 heap_id)
{
    struct UnkStruct_020851B8 *ptr = AllocFromHeap(heap_id, sizeof(struct UnkStruct_020851B8));
    __builtin__clear(ptr, sizeof(struct UnkStruct_020851B8));
    return ptr;
}
```

**Review by PikalaxALT — CHANGES_REQUESTED** [review]

**Review by PikalaxALT — APPROVED** [review]

**Pull request merged**